packages/uhk-web/src/app/components/macro/action-editor/tab/command/macro-command-editor.component.ts

```diff
@@ -163,10 +167,42 @@ export class MacroCommandEditorComponent implements AfterViewInit, ControlValueA
         this.removeNonAsciiCharachters();
     }
 
+    private setLanguageProperties(editor: MonacoStandaloneCodeEditor) {
+        const languageId = 'uhkScript';
+
+        if (!MacroCommandEditorComponent.completionRegistered) {
```

Comment:
This static variable is not really good for manage new language service registration, because if someone upgrade the firmware we have to apply the new language service.

We have to `registerCompletionItemProvider` and `setMonarchTokensProvider`.
We could register these new providers as we register the new themes in the `MonacoEditorCustomThemeService`. The difference is we have to re-register the providers every time when the configurations changed.

We can create a new `MonacoEditorUhkLanguageService` this service has to subscribe the `monacoLoaderService.isMonacoLoaded$` and when the new `getLanguageServiceConfig`. To combine 2 observers we could use [combineLatest](https://rxjs.dev/api/index/function/combineLatest) function.

How to create the `getLanguageServiceConfig` observerable?
We can extend the `packages/uhk-web/src/app/store/reducers/smart-macro-doc.reducer.ts` or create a new reducer. to be continue tomorrow

Reply:
The Agent has 2 main "process" the electron thread and the browser thread. The 2 world communicate with each other via electron built in [inter-process communication](https://www.electronjs.org/docs/latest/tutorial/ipc)

The `packages/uhk-common/src/util/ipcEvents.ts` file contains all of the ipc events that Agent uses. I tried to organise them. When the main module reply to an event in async mode I used the `Reply` suffix.

Every time when the user opens the smart macro document panel. 
- The renderer process dispatches the `SmartMacroDoc.downloadDocumentation` event.
- The main process checks the documentation for the firmware is available or not. if not then download it from the server. 

This moment is too late to load the auto-complete md file, because maybe the user start using the smart macro editor before opens the smart macro doc panel.
I think everytime when Agents starts or updates the firmware then have to check the smart macro document is available or not. If it not available, then have to download it and after provide the auto complete markdown file to the renderer process.

If we use the web version of the Agent then when Agents start we have to download the auto complete markdown file from the Agent repo. 

To achieve this functionality we have to refactor many small things. I don't know should I able to write every step in high level. My assumption is to provide a really step-by-step guide I have to implement the whole code in my mind. I am afraid if I forget something maybe I create a bigger problem for you than guidance.

I think faster if I finish the PR. Unfortunately, my availability will very limited in the next 3 weeks.
I am happy to help you to learn Agent coding, but I think we have to start with a smaller issue if you don't mind.

Reply:
> I think faster if I finish the PR.

Allright, it is yours then.

> Unfortunately, my availability will very limited in the next 3 weeks.

No worries ;)